generic-ec/src/multiscalar.rs

```diff
@@ -0,0 +1,404 @@
+//! multiplication is:
+//!
+//! $$\text{cost} = \log_2 s \cdot D + \frac{1}{2} \log_2 s \cdot A$$
+//!
+//! i.e. it does $\log_2 s$ doublings and $\frac{1}{2} \log_2 s$ additions. Naive
+//! multiscalar multiplication algorithm just computes $s_i P_i$ individually
+//! and then sums them. Total cost is:
+//!
+//! $$\text{cost} = n A + n (\log_2 s \cdot D + \frac{1}{2} \log_2 s \cdot A)$$
+//!
+//! Typically, we're working with scalars of 256 bits size. You can easily calculate
+//! that, for instance, even for $n = 3$, the total cost is $768 D + 387 A$. However,
+//! for the same $n$, Straus algorithm has $\text{cost} = 300 D + 192 A$ which
+//! is already significantly faster than naive algorithm.
+//!
+//! You can see the perfromance comparison for different algorithms on secp256k1 curve
+//! on the plot below:
+//!
+#![doc = include_str!("../perf/multiscalar/secp256k1.svg")]
+//!
+//! ## How to use it
+//! In most cases, all you need is [`Scalar::multiscalar_mul`] which defaults
+//! to the most efficient available algorithm, similarly to [`struct@Default`].
+//!
+//! Alternatively, if you need to use a specific algorithm, this module provides
+//! [`Straus`] and [`Pippenger`].
+
+use core::iter;
+
+#[cfg(feature = "alloc")]
+use alloc::vec::Vec;
+
+use crate::{Curve, Point, Radix16Iter, Scalar};
+
+/// Multiscalar multiplication algorithm
+///
+/// See [module-level docs](self) for motivation and list of provided algorithms.
+pub trait MultiscalarMul<E: Curve> {
+    /// Performs multiscalar multiplication
+    ///
+    /// Takes iterator of pairs `(scalar, point)`. Returns sum of `scalar * point`.
+    fn multiscalar_mul<S, P>(scalar_points: impl IntoIterator<Item = (S, P)>) -> Point<E>
+    where
+        S: AsRef<Scalar<E>>,
+        P: AsRef<Point<E>>;
+}
+
+/// Defaults to the most efficient multiscalar multiplication algorithm
+///
+/// When `alloc` feature is off, it always falls back to [`Naive`] implementation.
+///
+/// When `alloc` feature is on, it chooses the algorithm based on size of input `n`:
+/// * [`Straus`] when `n < 50`
+/// * [`Pippenger`] otherwise
+///
+/// It may be more convenient to use [`Scalar::multiscalar_mul`] which is an alias
+/// to `Default`.
+pub struct Default;
+
+#[cfg(not(feature = "alloc"))]
+impl<E: Curve> MultiscalarMul<E> for Default {
+    fn multiscalar_mul<S, P>(scalar_points: impl IntoIterator<Item = (S, P)>) -> Point<E>
+    where
+        S: AsRef<Scalar<E>>,
+        P: AsRef<Point<E>>,
+    {
+        Naive::multiscalar_mul(scalar_points)
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl<E: Curve> MultiscalarMul<E> for Default {
+    fn multiscalar_mul<S, P>(scalar_points: impl IntoIterator<Item = (S, P)>) -> Point<E>
+    where
+        S: AsRef<Scalar<E>>,
+        P: AsRef<Point<E>>,
+    {
+        let (mut scalars, points): (Vec<Radix16Iter<E>>, Vec<Point<E>>) = scalar_points
+            .into_iter()
+            .map(|(scalar, point)| (scalar.as_ref().as_radix16_be(), *point.as_ref()))
```

Comment:
Kind if stupid, but how do we guarantee that scalars are all the same length? I don't see us padding them with zeroes, and I don't see the iterator impl adding zeroes too

Reply:
`Scalar::to_{be|le}_bytes` always returns scalar zero-padded if needed